src/osm/GeometryHandler.cpp

```diff
@@ -61,137 +79,402 @@ osm2rdf::osm::GeometryHandler<W>::GeometryHandler(
-         std::to_string(innerId) +
-         ","
-         "\"inner_type\":\"" +
-         std::string(innerType) +
-         "\","
-         "\"duration_ns\":" +
-         std::to_string(durationNS.count()) +
-         ","
-         "\"result\":" +
-         (result ? "true" : "false") +
-         ""
-         "},\n";
-}
-
-// ____________________________________________________________________________
-template <typename W>
-void osm2rdf::osm::GeometryHandler<W>::area(const osm2rdf::osm::Area& area) {
-  auto geom = area.geom();
-  if (_config.simplifyGeometries > 0) {
-    geom = simplifyGeometry(geom);
+GeometryHandler<W>::~GeometryHandler() = default;
+
+// ____________________________________________________________________________
+template <typename W>
+void GeometryHandler<W>::relation(const Relation& rel) {
+  // careful, tags is returned via copy (why?)
+  const auto& tags = rel.tags();
+  auto typeTag = tags.find("type");
+  if (typeTag != tags.end() &&
+      (typeTag->second == "boundary" || typeTag->second == "multipolygon")) {
+    for (const auto& member : rel.members()) {
+      if (member.type() == RelationMemberType::WAY &&
+          (member.role() == "outer" || member.role() == "inner")) {
+#pragma omp critical
+        {
+          _areaBorderWaysIndex[member.id()].push_back(
+              {rel.id(), member.role() == "inner"});
+          std::sort(_areaBorderWaysIndex[member.id()].begin(),
+                    _areaBorderWaysIndex[member.id()].end(), MemberRelCmp());
+        }
+      }
+    }
+  }
+}
+
+// ____________________________________________________________________________
+template <typename W>
+void GeometryHandler<W>::area(const Area& area) {
+  // inner simplified geom, empty by default
+  osm2rdf::geometry::Area innerGeom;
+
+  // outer simplified geom, empty by default
+  osm2rdf::geometry::Area outerGeom;
+
+  // the convex hull of the geometry
+  osm2rdf::geometry::Polygon convexHull;
+
+  const auto& geom = simplifyGeometry(area.geom());
+
+  size_t totPoints = 0;
+  const size_t MIN_CUTOUT_POINTS = 10000;
+
+  for (const auto& poly : geom) {
+    totPoints += poly.outer().size();
+  }
+
+  if (!_config.dontUseInnerOuterGeoms && (area.hasName() || !area.fromWay())) {
+    innerGeom = simplifiedArea(geom, true);
+    outerGeom = simplifiedArea(geom, false);
+  }
+
+  if (!geom.empty()) {
+    boost::geometry::convex_hull(geom, convexHull);
+    if (convexHull.outer().size() >= geom.front().outer().size() / 2) {
+      // if we haven't gained anything significant in terms of number of anchor
+      // points, discard
+      convexHull = osm2rdf::geometry::Polygon();
+    }
+  }
+
+  std::vector<osm2rdf::geometry::Box> envelopes;
+  envelopes.push_back(area.envelope());
+  if (area.geom().size() > 1) {
+    for (const auto& polygon : area.geom()) {
+      osm2rdf::geometry::Box box;
+      boost::geometry::envelope(polygon, box);
+      envelopes.push_back(box);
+    }
+  } else {
```

Comment:
Maybe a small comment that this adds the envelope of every part of the a multipolygon with positive area. It's clear after looking at it, but it's not obious here why the distinction between `area.geom().size() >1` and `area.geom().size() <= 1` is needed. Is recalculating the Envelope that expensive?

Speaking of the `area.goem().size() <= 1` case: why is the evnelope added a second time? At the time of writing this seams to be the case that the first entry is always the bounding box of everything in the mulitpolygon and every subsequent entry for each part.

Is this done to simplify the code of later steps?

Reply:
I agree, I added a comment to clarify that. In general, calculating the bounding box is not expensive, but it felt wrong to iterate over thousands of anchor points just to generate a bounding box we already have :)

Yes, bounding boxes are stored two times for simple (non-multi) polygons to make later code easier.